codalab/worker/dependency_manager.py

```diff
@@ -263,130 +351,106 @@ def _cleanup(self):
             try:
-                path_to_remove = self._dependencies[dependency_key].path
-                self._paths.remove(path_to_remove)
+                path_to_remove = dependencies[dep_key].path
+                paths.remove(path_to_remove)
+                # Deletes dependency content from disk
                 remove_path(path_to_remove)
             except Exception:
                 pass
             finally:
-                del self._dependencies[dependency_key]
-                self._dependency_locks[dependency_key].release()
+                del dependencies[dep_key]
+                logger.info(f"Deleted dependency {dep_key}.")
 
     def has(self, dependency_key):
         """
-        Takes a DependencyKey
-        Returns true if the manager has processed this dependency
+        Takes a DependencyKey and returns true if the manager has processed this dependency
         """
-        with self._global_lock:
-            return dependency_key in self._dependencies
+        with self._state_lock:
+            dependencies: Dict[DependencyKey, DependencyState] = self._fetch_dependencies()
+            return dependency_key in dependencies
 
-    def get(self, uuid, dependency_key):
+    def get(self, uuid: str, dependency_key: DependencyKey) -> DependencyState:
         """
         Request the dependency for the run with uuid, registering uuid as a dependent of this dependency
         """
-        now = time.time()
-        if not self._acquire_if_exists(dependency_key):  # add dependency state if it does not exist
-            with self._global_lock:
-                self._dependency_locks[dependency_key] = threading.RLock()
-                self._dependency_locks[dependency_key].acquire()
-                self._dependencies[dependency_key] = DependencyState(
+        with self._state_lock:
+            dependencies, paths = self._fetch_state()
+
+            now = time.time()
+            # Add dependency state if it does not exist
+            if dependency_key not in dependencies:
+                # Determine dependency path
+                path = (
+                    os.path.join(dependency_key.parent_uuid, dependency_key.parent_path)
+                    if dependency_key.parent_path
+                    else dependency_key.parent_uuid
+                )
+                path = path.replace(os.path.sep, '_')
+
+                # Normalize the path for the dependency by replacing / with _, avoiding conflicts
+                while path in paths:
+                    path = path + '_'
+
+                # Add to the new path to the list of dependency state paths
+                paths.add(path)
+
+                dependencies[dependency_key] = DependencyState(
                     stage=DependencyStage.DOWNLOADING,
+                    downloading_by=None,
                     dependency_key=dependency_key,
-                    path=self._assign_path(dependency_key),
+                    path=path,
```

Comment:
Use `self._assign_path` instead for readability

```suggestion
                    path=self._assign_path(dependency_key, dependencies, paths),
```